pandas/io/parsers.py

```diff
@@ -1284,48 +1285,83 @@ def _agg_index(self, index, try_parse_dates=True):
                     col_na_values, col_na_fvalues = _get_na_values(
                         col_name, self.na_values, self.na_fvalues)
 
-            arr, _ = self._convert_types(arr, col_na_values | col_na_fvalues)
+            arr, _ = self._infer_types(arr, col_na_values | col_na_fvalues)
             arrays.append(arr)
 
         index = MultiIndex.from_arrays(arrays, names=self.index_names)
 
         return index
 
+    def _apply_converter(self, values, conv_f, na_values, col_na_values,
+                         col_na_fvalues):
+        """ apply converter function to values, respecting NAs """
+        try:
+            values = lib.map_infer(values, conv_f)
+        except ValueError:
+            mask = lib.ismember(values, na_values).view(np.uint8)
+            values = lib.map_infer_mask(values, conv_f, mask)
+
+        cvals, na_count = self._infer_types(
+            values, set(col_na_values) | col_na_fvalues,
+            try_num_bool=False)
+        return cvals, na_count
+
     def _convert_to_ndarrays(self, dct, na_values, na_fvalues, verbose=False,
-                             converters=None):
+                             converters=None, dtypes=None):
         result = {}
         for c, values in compat.iteritems(dct):
             conv_f = None if converters is None else converters.get(c, None)
+            if isinstance(dtypes, dict):
+                cast_type = dtypes.get(c, None)
+            else:
+                # single dtype or None
+                cast_type = dtypes
 
             if self.na_filter:
                 col_na_values, col_na_fvalues = _get_na_values(
                     c, na_values, na_fvalues)
             else:
                 col_na_values, col_na_fvalues = set(), set()
 
-            coerce_type = True
-            if conv_f is not None:
-                try:
-                    values = lib.map_infer(values, conv_f)
-                except ValueError:
-                    mask = lib.ismember(values, na_values).view(np.uint8)
-                    values = lib.map_infer_mask(values, conv_f, mask)
-                coerce_type = False
+            if conv_f is not None and cast_type is None:
+                # if type is not specified, apply the conversion first, without
+                # inference
+                cvals, na_count = self._apply_converter(
+                    values, conv_f, na_values,
+                    col_na_values, col_na_fvalues)
+            else:
+                try_num_bool = True
```

Comment:
Why not just this:
~~~python
try_num_bool = cast_type and is_object_dtype(cast_type)
~~~